include/crypto/event.h

```diff
@@ -0,0 +1,378 @@
+ * dynamically allocated.
+ */
+struct ossl_event_st;
+typedef int ossl_event_destructor_fn(struct ossl_event_st *event);
+
+/*
+ * Generic callback function type.  This is used for subscribers and filters
+ * alike, as seen in the different functions where this type is used.
+ */
+typedef int ossl_event_callback_fn(struct ossl_event_st *event,
+                                   void *callback_data);
+
+struct ossl_event_st {
+    /*
+     * Mandatory data.
+     *
+     * These fields are considered immutable.  However, the contents that
+     * |ctx| points at may be mutable.
+     */
+
+    /* A numeric event identity, defined by the using functionality */
+    uint32_t type;
+    /* A context relevant for the using functionality */
+    void *ctx;
+
+    /*
+     * Optional data (may be NULL).
+     *
+     * These fields are considered mutable, in so far that an event handler
+     * may "steal" them, and replace them with NULL.
+     */
+
+    /* Identifying material */
+    const void *identifiers;
+    /* Data carried by the event */
+    size_t payload_size;
+    void *payload;
+
+    /*
+     * A destructor for |identifiers| and |payload|, if necessary.
+     *
+     * This field is considered immutable until the event is destroyed.
+     */
+    ossl_event_destructor_fn *destructor;
+
+    /*
+     * Flags, only relevant for the event functionality itself.
+     */
+    unsigned int flag_dynamic : 1;
+};
+
+/*
+ * The event queue is expected to be very long lived, even thought it may be
+ * dynamically allocated.  Subsribers are expected to return values that
+ * express if the event it received was used or not.
+ */
+enum ossl_event_subscriber_returns {
+    /* The subscriber encountered an error */
+    ossl_event_error = -1,
+    /* The subscriber didn't use this event */
+    ossl_event_pass = 0,
+    /* The subscriber used this event */
+    ossl_event_handled = 1,
+};
+
+/*
+ * OSSL_EVENT_SUBSCRIBER_CLOSURE implements a closure simile, speciflcally
+ * for our event callbacks, to carry along the callback specific data pointer.
+ */
+typedef struct ossl_event_subscriber_closure_st {
+    ossl_event_callback_fn *cb;
+    void *cbarg;
+} OSSL_EVENT_SUBSCRIBER_CLOSURE;
+
+/*
+ * Opaque definition, to allow an optimal internal implementation without
+ * having to change anything else.
+ * This structure is expected to have the semantics of an array of stacks
+ * of events, and should only be affected by event addition, event removal
+ * and event dispatching, all of is internal code in crypto/event/event.c.
+ *
+ * Each element in this implied array represent a priority level, and each
+ * stack (element in this array) simply stores the events for its priority
+ * level in a FIFO manner; events get pushed in and shifted out.
+ */
+struct ossl_events_stacks_st;
+
+struct ossl_event_queue_method_st;
+struct ossl_event_queue_st {
+    const struct ossl_event_queue_method_st *method;
```

Comment:
This is serious overkill for a private internal API.  We shouldn't need a method structure.

How many different kinds of event queues are you anticipating?  I'm thinking exactly one.  We can add methods support if there is ever a second.

Keep it simple.

Reply:
I've imagined multiple queues with different characteristics,  If that's overkill or not, I cannot say.

Reply:
It's overkill :)

Reply:
I can't think why we would need multiple queues with different characteristics. If we need that in the future then we can add it. So, I tend to agree with @paulidale. Let's not over-engineer it.

Reply:
I don't actually believe that it's over-engineered, but hey, what do I know...  so yeah ok, I can minimize this...  (I made a branch to preserve the over-engineered bits, https://github.com/levitte/openssl/tree/event-queue-design@@over-engineered, just in case)